src/backend/distributed/planner/relation_restriction_equivalence.c

```diff
@@ -1052,3 +1259,93 @@ AttributeEquivalancesAreEqual(AttributeEquivalenceClass *firstAttributeEquivalan
 
 	return true;
 }
+
+
+/*
+ *  TopLevelUnionQuery gets a queryTree and returns true if top level query
+ * contains a UNION set operation. Note that the function allows upper level
+ * queries with a single range table element. In other words, we allow top
+ * level unions being wrapped into aggregations queries and/or simple projection
+ * queries that only selects some fields from the lower level queries.
+ *
+ * If there exists joins before the set operations, the function returns false.
+ * Similarly, if the query does not contain any union set operations, the
+ * function returns false.
+ */
+static bool
+TopLevelUnionQuery(Query *queryTree)
+{
+	List *rangeTableList = queryTree->rtable;
+	Node *setOperations = queryTree->setOperations;
+	RangeTblEntry *rangeTableEntry = NULL;
+	Query *subqueryTree = NULL;
+
+	/* don't allow joines on top of unions */
+	if (list_length(rangeTableList) > 1)
+	{
+		return false;
+	}
+
+	rangeTableEntry = linitial(rangeTableList);
+	if (rangeTableEntry->rtekind != RTE_SUBQUERY)
+	{
+		return false;
+	}
+
+	subqueryTree = rangeTableEntry->subquery;
+	setOperations = subqueryTree->setOperations;
+	if (setOperations != NULL)
+	{
+		SetOperationStmt *setOperationStatement = (SetOperationStmt *) setOperations;
+
+		if (setOperationStatement->op != SETOP_UNION)
+		{
+			return false;
+		}
+
+		return true;
+	}
+
+	return TopLevelUnionQuery(subqueryTree);
```

Comment:
I did not like recursion here.  I think this is used to traverse query tree to find out the first query where there is either set operation or join. 

Can you move search logic to another walker like function and remove  recursion here ?